interface.rsh

```diff
@@ -1,42 +1,191 @@
-  .timeout(relativeTime(ttl), () => {
-    Anybody.publish()
-    transfer(balance()).to(addr)
-    commit();
-    exit();
-  })
+  Manager.publish(tokenAmount, price)
+    .pay([amt, [tokenAmount, token]])
+    .check(() => {
+      check(tokenAmount > 0, "tokenAmount must be greater than 0");
+      check(price > 0, "price must be greater than 0");
+    })
+    .timeout(relativeTime(ttl), () => {
+      Relay.only(() => {
+        const rAddr = this;
+      });
+      Relay.publish(rAddr);
+      transfer([[getUntrackedFunds(token), token]]).to(rAddr);
+      commit();
+      exit();
+    });
   transfer(amt).to(addr);
+  Manager.interact.signal();
+
+  const initialState = {
+    manager: Manager,
+    token,
+    tokenAmount,
+    price,
+    closed: false,
+  };
+
+  const [s] = parallelReduce([initialState])
+    .define(() => {
+      v.state.set(State.fromObject(s));
+    })
+    // TOKEN BALANCE
+    .invariant(
+      implies(!s.closed, balance(token) == s.tokenAmount),
+      "token balance accurate before close"
+    )
+    .invariant(
+      implies(s.closed, balance(token) == 0),
+      "token balance accurate after close"
+    )
+    // BALANCE
+    .invariant(balance() == 0, "balance accurate")
+    .while(!s.closed)
+    // api: update
+    //  - update price
+    .api_(a.update, (msg) => {
+      check(msg > 0, "price must be greater than 0");
+      return [
+        (k) => {
+          k(null);
+          return [
+            {
+              ...s,
+              price: msg,
+            },
+          ];
+        },
+      ];
+    })
+    // api: grant
+    //  - asign another account as manager
+    .api_(a.grant, (msg) => {
+      check(this === s.manager);
+      return [
+        (k) => {
+          k(null);
+          return [
+            {
+              ...s,
+              manager: msg,
+            },
+          ];
+        },
+      ];
+    })
+    // api: buy
+    //  - buy token
+    .api_(a.buy, (msg) => {
+      check(msg <= s.tokenAmount, "not enough tokens");
+      return [
+        msg * price,
+        (k) => {
+          k(null);
+          transfer(msg * price).to(s.manager);
+          transfer(msg, token).to(this);
+          return [
+            {
+              ...s,
+              tokenAmount: s.tokenAmount - msg,
+            },
+          ];
+        },
+      ];
+    })
+    // api: close
+    //  - close contract
+    .api_(a.close, () => {
+      check(this == s.manager);
+      return [
+        (k) => {
+          k(null);
+          transfer([[s.tokenAmount, token]]).to(s.manager);
+          return [
+            {
+              ...s,
+              closed: true,
+              tokenAmount: 0,
+            },
+          ];
+        },
+      ];
+    })
+    .timeout(false);
   commit();
-  Relay.publish();
-  transfer(balance()).to(addr)
+  Relay.only(() => {
+    const rAddr = this;
+  });
+  Relay.publish(rAddr);
+  transfer([[getUntrackedFunds(token), token]]).to(rAddr);
```

Comment:
There's no guarantee that this is the same relay that ever did anything useful and I continue to disagree that "relay"s do anything useful.

Reply:
Got it. @jeapostrophe would it matter if it were a Participant?

Reply:
A participant would be set after the first time, so I think it would be better that way